internal/partition/partition.go

```diff
@@ -0,0 +1,220 @@
+func NewPartitionTable(volume *gadget.Volume, sectorSize uint64, imgSize uint64) Table {
+	if volume.Schema == SchemaMBR {
+		return &MBRTable{
+			diskSize: imgSize,
+			concreteTable: &mbr.Table{
+				LogicalSectorSize:  int(sectorSize),
+				PhysicalSectorSize: int(sectorSize),
+			},
+		}
+	}
+	return &GPTTable{
+		diskSize: imgSize,
+		concreteTable: &gpt.Table{
+			LogicalSectorSize:  int(sectorSize),
+			PhysicalSectorSize: int(sectorSize),
+			ProtectiveMBR:      true,
+		},
+	}
+}
+
+// GeneratePartitionTable prepares the partition table for structures in a volume and
+// returns it with the partition number of the root partition.
+func GeneratePartitionTable(volume *gadget.Volume, sectorSize uint64, imgSize uint64, isSeeded bool) (partition.Table, int, error) {
+	partitionNumber, rootfsPartitionNumber := 1, -1
+	partitionTable := NewPartitionTable(volume, sectorSize, imgSize)
+	onDisk := gadget.OnDiskStructsFromGadget(volume)
+
+	for i := range volume.Structure {
+		structure := &volume.Structure[i]
+		if !structure.IsPartition() || helper.ShouldSkipStructure(structure, isSeeded) {
+			continue
+		}
+
+		// Record the actual partition number of the root partition, as it
+		// might be useful for certain operations (like updating the bootloader)
+		if helper.IsRootfsStructure(structure) { //nolint:gosec,G301
+			rootfsPartitionNumber = partitionNumber
+		}
+
+		structurePair := &gadget.OnDiskAndGadgetStructurePair{
+			DiskStructure:   onDisk[structure.YamlIndex],
+			GadgetStructure: &volume.Structure[i],
+		}
+
+		structureType := getStructureType(structure, volume.Schema)
+		err := partitionTable.AddPartition(structurePair, structureType)
+		if err != nil {
+			return nil, rootfsPartitionNumber, err
+		}
+
+		partitionNumber++
+	}
+
+	return partitionTable.GetConcreteTable(), rootfsPartitionNumber, nil
+}
+
+// getStructureType extracts the structure type from the structure.Type considering
+// the schema
+func getStructureType(structure *gadget.VolumeStructure, schema string) string {
+	structureType := structure.Type
+	// Check for hybrid MBR/GPT
+	if !strings.Contains(structure.Type, ",") {
+		return structureType
+	}
+
+	types := strings.Split(structure.Type, ",")
+	structureType = types[0]
+
+	if schema == SchemaGPT {
+		structureType = types[1]
+	}
+
+	return structureType
+}
+
+type MBRTable struct {
+	concreteTable *mbr.Table
+	diskSize      uint64
+}
+
+func (t *MBRTable) AddPartition(structurePair *gadget.OnDiskAndGadgetStructurePair, structureType string) error {
+	// mbr.Type is a byte. snapd has already verified that this string
+	// is exactly two chars, so we can safely parse those two chars to a byte
+	partitionType, _ := strconv.ParseUint(structureType, 16, 8) // nolint: errcheck
+
+	t.concreteTable.Partitions = append(t.concreteTable.Partitions, &mbr.Partition{
+		Start:    uint32(math.Ceil(float64(structurePair.DiskStructure.StartOffset) / float64(t.concreteTable.LogicalSectorSize))),
```

Comment:
It's very much a tangent to everything else going on but I would prefer these conversions between bytes and sectors stay in the integer domain.

Reply:
I was more tempted to say in `uint64` to be sure this is always positive but I agree. Sadly we have to deal with the fact that the go-diskfs lib we use expects uint32 (for MBR partitions), uint64 (for GPT partitions) and on the snapd size they have a specific type `quantity.Size` (which is a type based on uint64). I tried to juggle with all this at best by casting as late as possible but if you think this could be improved let me know.

Reply:
Eh I wasn't clear here. I meant integer operations rather than doing things as floats, so things like `(offset + sector_size - 1) / sector_size` to go from bytes to sectors.